### PR TITLE
Initial concept for new accordion

### LINKED ISSUE #1703: Missing parent role="tablist", ce_accordionStart.html5
In Contao 4.4.47 the Webdeveloper audit shows an accessibility error caused by a missing parent role="tablist"  in ce_accordionStart.html5. I fixed it for me. I just created a new ce_accordionStart.html5 template with the missing role:

`<section role="tablist" class="<?= $this->class ?> ce_accordion block"<?= $this->cssID ?><?php if ($this->style): ?> style="<?= $this->style ?>"<?php endif; ?>>`

Is that already implemented in higher Contao Versions after 4.4.47? 

Issue related Link: https://www.w3.org/TR/wai-aria-1.1/#tab

### PR DESCRIPTION
**Pull request by fritzmg**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #1703 
| Docs PR or issue | -

This is a draft PR showing how we could implement different accordion libraries, which require a wrapper around all accordion elements (which the current jQuery UI solution does not). The start and stop content elements simply make another database query checking whether they are preceded or followed by  a stop or start element respectively and thus only add the wrapper accordingly.

### CODE REVIEW

**Comment by m-vo** [comment]
What about nested structures?

**Comment by fritzmg** [comment]
> What about nested structures?

The solution would work with nested structures, however, we are deciding against it, as it would be too complex to consider all cases, especially protected elements, or elements that are only shown to guests.

**Comment by m-vo** [comment]
Maybe we should give real nested elements an attempt first and use this to solve the accordion problem as well.

**Comment by aschempp** [comment]
I still don't like the idea of adding a wrapper. I know I've built cases (years ago) where the accordions were spread across the whole site (e.g. opening an according in the main column also collapsed the ones in the right column). 

Also, do I assume correctly that this will break any existing custom accordion templates? Or will it just be a "new way" of using accordions?

**Comment by fritzmg** [comment]
> I know I've built cases (years ago) where the accordions were spread across the whole site (e.g. opening an according in the main column also collapsed the ones in the right column).

imho that's not desirable behaviour. If I have multiple groups of accordions on a page, I would expect them to act independently from each other.

No matter which solution we have in the future, having a dedicated wrapper around all accordion is the way to go in my opinion.

> Also, do I assume correctly that this will break any existing custom accordion templates? Or will it just be a "new way" of using accordions?

There is no BC promise for templates iirc, but in any case, the accordions would still work in the new version, if you have a custom template. They just won't be grouped together.

**Comment by MacKP** [comment]
I agree with @fritzmg here (Wrapper is always useful in such case and adds a meaning to it. Other Cases like @aschempp mentioned are custom things and are not really accessible)
But I would prefer a solution where the Button is set with JS. Without JS the Button makes no sence here.

**Comment by gmpf** [comment]
> The start and stop content elements simply make another database query checking whether they are preceded or followed by a stop or start element respectively

Maybe I'm missing something, but wouldn't the full implementation of this have to become a bit more complex? The queries would have to take `invisible`, `start`, and `stop` into account, like the `*PublishedBy*` methods in `ContentModel`, and there would also need to be a `Controller::isVisibleElement()` check. The latter would then necessitate further logic for retrieving the actual visible adjacent element.

**Comment by fritzmg** [comment]
@gmpf yes, see https://github.com/contao/contao/pull/2301#issuecomment-692961670

> however, we are deciding against it, as it would be too complex to consider all cases, especially protected elements, or elements that are only shown to guests.

**Comment by gmpf** [comment]
@fritzmg I see, I initially read that quote as being exclusively in reference to taking nested structures into account. Is my understanding correct then that given this structure in the back end:
```
1:ContentAccordionStart
2:	Content
3:	Content
4:ContentAccordionStop
5:invisible Content
6:ContentAccordionStart
7:	Content
8:	Content
9:ContentAccordionStop
```
we would end up with two individually-wrapped accordion elements in the front end:
```
<wrapper>1,2,3,4</wrapper>
<wrapper>6,7,8,9</wrapper>
```
as opposed to one wrapper:
```
<wrapper>1,2,3,4,6,7,8,9</wrapper>
```
and that behaviour would be an accepted limitation?

**Comment by fritzmg** [comment]
@gmpf do you mean in the current implementation or in the implementation of this draft? In the implementation of this draft, this would create two accordion groups, due to the limitations mentioned in https://github.com/contao/contao/pull/2301#issuecomment-692961670. Unless the content element is simply "invisible" due to `tl_content.invisible`, then it would be one group.

In the current implementation, accordions aren't grouped at all.

> and that behaviour would be an accepted limitation?

No, it wouldn't be an accepted limitation, hence the following comment: https://github.com/contao/contao/pull/2301#issuecomment-692961670

> however, we are deciding against it, as it would be too complex to consider all cases, especially protected elements, or elements that are only shown to guests.

;)

There is no new proposal as of yet.

**Review by aschempp — COMMENTED** [review]
as already said, I'm not a fan of the wrappers. However, **if** we really need them, the should be generated on-the-fly in the front end and we must not try in the back end. It will never work consistently in the back end …

**Comment by fritzmg** [comment]
Why? We have wrappers for everything else. Accordions should have wrappers too - they are currently the odd one.

**Comment by asaage** [comment]
> ... which require a wrapper around all accordion elements (which the current jQuery UI solution does not). 

Honestly i dont't think that is 100% correct.
The [example jQuery UI Accordion](https://jqueryui.com/accordion/) uses a wrapper and the wrapper should be used to initialize the accordion (which Contao does not do).
https://github.com/contao/contao/blob/master/core-bundle/src/Resources/contao/templates/jquery/j_accordion.html5#L5

We already have Wrapper-elements (Accordion (Wrapper start)/(Wrapper stop)) but the naming is misleading:
they dont wrap around an accordion but around a single collapsible item.

image

I find myself messing with the accordion templates in almost every installation because i absolutely don't like how something in ```#main``` affects something completely different in  ```#left``` or ```#right```

**Comment by stefansl** [comment]
I like the approach using handorgel for a new Contao accordion. Since this is a new template (js_accordion instead of j_accordion) this is imho template wise no BC break.
But we should keep our Contao classnames ce_accordion, ce_accordion_header and accordion - it's easier to switch to the new js template

**Comment by MacKP** [comment]
@stefansl hadorgel is an nice peace of code and it's realy lightweight. The only thing I don't like about it is the markup with buttons (without JS, there is no content and you only get the buttons). Because of this I would recomend a JS just for changing the Markup from headings and content to the Markup with Buttons. Then you get just good old plain semantic HTML without JS and everything works fine :-)

**Comment by stefansl** [comment]
@MacKP you're right. Maybe it's a good idea to have a working pure CSS solution with a little bit of JS on top for extra options like changing classnames, collapse all/none, etc.

**Comment by MacKP** [comment]
> @MacKP you're right. Maybe it's a good idea to have a working pure CSS solution with a little bit of JS on top for extra options like changing classnames, collapse all/none, etc.

@stefansl just skip a CSS solution for Accordions. It should just output a heading with the content. All CSS only Sollutions are just providing more problems than solutions ;-)

**Comment by aschempp** [comment]
HTML `detail` and `summary` sound like an alternative to me? 🙃 
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details

**Comment by MacKP** [comment]
> HTML `detail` and `summary` sound like an alternative to me? upside_down_face
> https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details

It's not a good alternative  here. See -> [https://daverupert.com/2019/12/why-details-is-not-an-accordion/](url)

**Comment by aschempp** [comment]
TLDR: The problem of `<details>` is that `<summary>` is a button and not a headline, which boils down to screenreader problems (e.g. navigating headlines). Doesn't that also disqualify Handorgel because it uses buttons for headlines? On the other hand, how does a screenreader know it can toggle a headline, I thought that's exactly what buttons are used for?

**Comment by MacKP** [comment]
As the example shows -> use a Button in the headline. That would be the best solution.  I am all in for native HTML. But the details and summary is not the best way to do it here. Just use a button in a headline. And without JavaScript its just a Headline and the content without extra Markup.

**Comment by aschempp** [comment]
which one do you think is better?
1. `<h1><button>this is the headline</button></h1>`
2. `<h1>this is the headline <button>toggle</button></h1>`?

**Comment by m-vo** [comment]
The problem really is that lots of screen readers are lacking behind years of progress the HTML spec has made… That is very unfortunate because the problem IMO are the screen readers not the web, but that does not help the people using them.

**Comment by MacKP** [comment]
> The problem really is that lots of screen readers are lacking behind years of progress the HTML spec has made… That is very unfortunate because the problem IMO are the screen readers not the web, but that does not help the people using them.

Not only that.. but summary is not allowed in an headline...

**Comment by MacKP** [comment]
> which one do you think is better?
> 
>     1. `<h1><button>this is the headline</button></h1>`
> 
>     2. `<h1>this is the headline <button>toggle</button></h1>`?

I like the first one. The second one might not be generic enough. But it always depends on the usecase xD
But most of the time I implenent it like the first one with aria-expanded.

**Comment by leofeyer** [comment]
Here is a good concept how to integrate Handorgel (in an easy way): https://www.infodrog.ch/de/kontakt.html

**Comment by MacKP** [comment]
The only changes would be:

- Use Classes from Contao
- Implementing the Button with JS

**Comment by aschempp** [comment]
@fritzmg are you planning to adjust this PR or should it be a separate one with my draft?

**Comment by fritzmg** [comment]
> @fritzmg are you planning to adjust this PR or should it be a separate one with my draft?

Sorry, never responded to this. Create a separate one with your draft please :)

**Comment by aschempp** [comment]
I guess this is obsolete with https://github.com/contao/contao/pull/5875 and https://github.com/contao/contao/pull/6137 now? 🙃

**Comment by fritzmg** [comment]
Not needed at all anymore with Contao 5.3 and its new Handorgel Nested Content Element implementation 👍
